Posts about Patterns (#16)
aseriousbiz/blog: _posts/2021/2021-11-03-how-cascadiajs-uses-abbot.md

---
title: "How CascadiaJS uses Abbot to power its live feed"
description: "CascadiaJS uses Abbot patterns to power its live feed. See how to use Patterns in this post"
tags: [patterns,cascadiajs,abbot]
excerpt_image: 
author:
    avatar: https://avatars.githubusercontent.com/u/279389?s=400&u=74e6e598a2bf9d9889fbb771c542c508bf270e36&v=4
    name: pmn
---

> Attendees to CascadiaJS saw updates from the Cascadia chat from the Cascadia Discord as soon as the messages were sent. 

Reviewer: This sentence is a bit awkward and hard to parse. Where did they see these updates? Maybe,

```suggestion
During the CascadiaJS conference, attendees could see live updates on screens around the conference with the latest chat messages from the Cascadia Discord. 
```

Or something like that?

> If you attended CascadiaJS 2021, you may have noticed live updates on the website that came from Cascadia's Discord server. 

Author: Agree, but I haven't been able to access it yet. May grab one tomorrow once it's live.

You could build something like this yourself. Start by [signing up for an Abbot account](https://ab.bot) and connecting it to your chat server (Abbot supports Slack, Discord, or Microsoft Teams). 

CascadiaJS sends every message that Abbot sees out to another system. Let’s build something that is a little less noisy. Once you're done, you’ll have all the tools you need to build anything you’d like.

Let's build a simple script that listens for people asking if a site is down (__“is www.aseriousbusiness.com down?”__). If Abbot is in the room when someone asks the question, it can try to connect to the server and let the user know if the site is up or down. 

Since this was done for a JavaScript conference, let's build our script using JavaScript. Log in to Abbot and create a new JavaScript skill called `website-checker`. Copy and paste the code below into your skill:

```javascript
  const axios = require('axios');

>   let website = bot.arguments; // bot.arguments contains the user-provided text
>   if (bot.isPatternMatch) {    // this will be true if it's the result of a pattern match. Otherwise, it's a chat message.
>     website = bot.tokenizedArguments[1].toString(); // Abbot tokenizes arguments to make it easier to ahandle them.
>   }

Reviewer: This is a stylistic choice, but if this were my code, I'd probably do this. :smile:

```suggestion
  // bot.arguments contains the user-provided text
  // bot.tokenizedArguments contains parsed arguments to make it easier to handle them.
  const website = bot.isPatternMatch ? bot.tokenizedArguments[1].value : bot.arguments;
```

  
  await axios.get(website)
    .then(function (response) {
        // handle success

>         bot.reply(`${bot.from} it's just you.. it looks okay to me...`)

Reviewer: ```suggestion
        await bot.reply(`${bot.from} it's just you.. it looks okay to me...`)
```

Author: For whatever reason, this doesn't work inside of the `await`ed `axios.get` (it complains about not being inside of an `async` function). Going to leave it as is because it's functioning correctly on test.

      })
      .catch(function (error) {
        // handle error

>         bot.reply(`it's not just you ${bot.from}, i got this error when i tried: \n ${error}`)

Reviewer: ```suggestion
        await bot.reply(`it's not just you ${bot.from}, i got this error when i tried: \n ${error}`)
```

      });
```

Test the skill in the console by entering the domain of a website you want to check. In the console, you will need to include `https://` in front of the domain name. 

Once you’ve verified that the skill works, add a Pattern by clicking the link that says “Manage skill patterns” on the left side of the screen. Click “Create Pattern”, and give the new pattern a descriptive name like “url pattern”. To save time, pick a pattern type of “Regular Expression” and use this regular expression: 

`is (<)?(http:\/\/|https:\/\/)?[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b([-a-zA-Z0-9()@:%_\+.~#?&//=]*)(|.*?)(>?) down`

This Pattern is made a little more complicated by all the ways chat platforms might send us the URL, for now you'll just have to trust that it works. You can see for yourself by using the built-in tester. Try a phrase like “is google.com down?” to make sure it matches.

> Once this is enabled, jump into any chat room that has Abbot in it, and try asking if a site is up or down. Go on, we’ll wait for you to try it! 

Reviewer 2: Great example, love it!
